File: app/src/main/java/idrisadetunmbi/githubusers/UsersContract.java
package idrisadetunmbi.githubusers;

import java.util.List;

import idrisadetunmbi.githubusers.models.GithubUser;

public interface UsersContract {
    interface View {
        void setProgressIndicator(boolean isLoading);

        void showUsers(List<GithubUser> users);

        void showUserDetailsUi(GithubUser user);

        boolean userIsOnline();

        void showNetworkUnavailabilitySnackbar();
    }

    interface ActionsListener {
        void loadUsers();

        void openDetailsView(GithubUser user);
    }

    interface Data {
        void getUsers(Callback<List<GithubUser>> callback);

        interface Callback<T> {
            void onLoaded(T data);
        }
    }
}

File: app/src/main/java/idrisadetunmbi/githubusers/UsersContract.kt
package idrisadetunmbi.githubusers

import idrisadetunmbi.githubusers.models.GithubUser

interface UsersContract {
    interface View {
        fun setProgressIndicator(isLoading: Boolean)

        fun showUsers(users: List<GithubUser>)

        fun showUserDetailsUi(user: GithubUser)

        fun deviceIsConnected(): Boolean

        fun showNetworkUnavailabilitySnackbar()
    }

    interface ActionsListener {
        fun loadUsers()

        fun openDetailsView(user: GithubUser)
    }

    interface Data {
        fun getUsers(callback: Callback<List<GithubUser>>)

        interface Callback<T> {
            fun onLoaded(data: T)
        }
    }
}
